src/frequenz/client/dispatch/_client.py

```diff
@@ -0,0 +1,101 @@
+# License: MIT
+# Copyright © 2024 Frequenz Energy-as-a-Service GmbH
+
+"""Dispatch API client for Python."""
+from dataclasses import dataclass
+from datetime import datetime
+
+import grpc
+from frequenz.api.dispatch.v1 import dispatch_pb2, dispatch_pb2_grpc
+from frequenz.sdk.microgrid.component import ComponentCategory
+from google.protobuf.timestamp_pb2 import Timestamp
+
+# from ._types import ForecastFeature, Forecasts, Location
+
+
+@dataclass
+class TimeIntervalFilter:
+    """Filter for a time interval."""
+
+    start_from: datetime
+    """Filter by start_time >= start_from."""
+
+    start_to: datetime
+    """Filter by start_time < start_to."""
+
+    end_from: datetime
+    """Filter by end_time >= end_from."""
+
+    end_to: datetime
+    """Filter by end_time < end_to."""
+
+
+class Client:
+    """Dispatch API client."""
+
+    def __init__(self, grpc_channel: grpc.aio.Channel, svc_addr: str) -> None:
+        """Initialize the client.
+
+        Args:
+            grpc_channel: gRPC channel to use for communication with the API.
+            svc_addr: Address of the service to connect to.
+        """
+        self._svc_addr = svc_addr
+        self._stub = dispatch_pb2_grpc.MicrogridDispatchServiceStub(grpc_channel)
+
+    # pylint: disable=too-many-arguments
+    async def list(
+        self,
+        microgrid_id: int,
+        component_selectors: list[list[int] | ComponentCategory] | None = None,
+        interval_filter: TimeIntervalFilter | None = None,
+        is_active: bool | None = None,
+        is_dry_run: bool | None = None,
+    ) -> None:
+        """List dispatches.
+
+        Args:
+            microgrid_id: The microgrid_id to list dispatches for.
+            component_selectors: optional, list of component ids or categories to filter by.
+            interval_filter: optional, filter by time interval.
+            is_active: optional, filter by is_active status.
+            is_dry_run: optional, filter by is_dry_run status.
+        """
+        time_interval = None
+
+        def to_timestamp(dt: datetime) -> Timestamp:
+            ts = Timestamp()
+            ts.FromDatetime(dt)
+            return ts
+
+        if interval_filter:
+            time_interval = dispatch_pb2.TimeIntervalFilter(
+                start_from=to_timestamp(interval_filter.start_from),
+                start_to=to_timestamp(interval_filter.start_to),
+                end_from=to_timestamp(interval_filter.end_from),
+                end_to=to_timestamp(interval_filter.end_to),
+            )
```

Comment:
The idea is that your `TimeIntervalFilter` wrapper class has `from_proto()` and `to_proto()` methods. For now in the SDK we only have `from_proto()` ([example](https://github.com/frequenz-floss/frequenz-sdk-python/blob/91821a3e2719024c0b2a114d6a11aa692d3429d6/src/frequenz/sdk/microgrid/component/_component_data.py#L85-L112)) but I guess it could make sense to have the other too.

That said, to be honest, I don't like this approach, as the idea is to hide the underlying protocol we are using in the client, the idea is that the client user doesn't need to know about protobuf. Since conversion only needs to happen in the client internally, I would prefer an approach where we provide conversion functions separately from the wrapper class, similar to what marshmallow does, or even dataclass when converting to `dict` or `tuple`.

Maybe this is something to discuss in the next API meeting (or even SDK meeting, as this is a Python thing more than a API thing).

FYI @frequenz-floss/python-sdk-team

Reply:
I am not sure for this particular type it's worth having all that, given that it's only used in this function and no-where outside, and the conversion is extremely simple

Reply:
Right, maybe I picked the wrong example.